High-availability guide restructuring

ryanemerson opened the pull request:
Closes #41585
Closes #30095

To build the website build this PR and then build the following keycloak-web branch (WIP): https://github.com/ryanemerson/keycloak-web/tree/multi_az_docs

ryanemerson committed:
Refactor high-availability guide to include both single and multi cluster architectures

Closes #30095
Closes #41585

Signed-off-by: Ryan Emerson <remerson@ibm.com>
Signed-off-by: Alexander Schwartz <aschwart@redhat.com>
Co-authored-by: Alexander Schwartz <aschwart@redhat.com>
Signed-off-by: Ryan Emerson <remerson@ibm.com>

pruivo commented on docs/guides/high-availability/multi-cluster/concepts-threads.adoc:
```diff
@@ -0,0 +1,14 @@
+This section is intended when you want to understand the considerations and best practices on how to configure thread pools connection pools for {project_name}.
+For a configuration where this is applied, visit <@links.ha id="multi-cluster-deploy-keycloak-kubernetes" />.
+
+<#include "/high-availability/partials/concepts/threads.adoc" />
```
JGroups thread pool is mentioned here while it is not used in multi-cluster deployments.

ryanemerson commented on docs/guides/high-availability/single-cluster/introduction.adoc:
```diff
@@ -0,0 +1,90 @@
+either ROSA HCP or ROSA classic.
+</@profile.ifProduct>
+<@profile.ifCommunity>
+using ROSA HCP.
```
Not sure, this is the same as what was in the multi-site guide.

ryanemerson commented on docs/guides/high-availability/single-cluster/concepts.adoc:
```diff
@@ -0,0 +1,161 @@
+| No data loss
+| Seconds to minutes
+
+|===
```
I originally thought the same, but this section is "Failures which this setup can survive" and we definitely can't survive a DB crash :slightly_smiling_face:

ryanemerson commented on docs/guides/high-availability/single-cluster/concepts.adoc:
```diff
@@ -0,0 +1,161 @@
+| No data loss^3^
+| Seconds to minutes (depending on the database)
+
+| Connectivity {jdgserver_name}
```
See previous reply.

ahus1 commented on docs/guides/high-availability/partials/concepts/perf_recommendations.adoc:
```diff
@@ -0,0 +1,49 @@
++
+{project_name} spends most of the CPU time hashing the password provided by the user, and it is proportional to the number of hash iterations.
+
+* For each 120 client credential grants per second, 1 vCPU to the cluster (tested with up to 2000 per second).^*^
```
That's a poor mans footnote.

ryanemerson replied:
No idea, I just moved this existing content, I'll remove it.

ryanemerson replied:
Actually, it's provided later in the docs:

```quote
For non-ARM CPU architectures on AWS (c7i/c7a vs. c7g) we found ...
```

ryanemerson commented on docs/guides/high-availability/single-cluster/deploy-keycloak.adoc:
```diff
@@ -0,0 +1,94 @@
+[#single-cluster-deploy-keycloak-prerequisites]
+== Prerequisites
+
+* OpenShift or Kubernetes cluster deployed across multiple availability-zones with a worker-pool configured for each.
```
In the existing multi-site guide we say "OpenShift or Kubernetes cluster running."

ryanemerson commented on docs/guides/high-availability/single-cluster/deploy-keycloak.adoc:
```diff
@@ -0,0 +1,94 @@
+
+* OpenShift or Kubernetes cluster deployed across multiple availability-zones with a worker-pool configured for each.
+* Understanding of a <@links.operator id="basic-deployment" /> of {project_name} with the {project_name} Operator.
+* Aurora AWS database deployed using the <@links.ha id="single-cluster-deploy-aurora" /> {section}.
```
It seems we had this inconsistency already. I have updated all occurences to use "AWS Aurora" as that was more prevalent.

ryanemerson commented on docs/guides/high-availability/single-cluster/deploy-keycloak.adoc:
```diff
@@ -0,0 +1,94 @@
+<#include "/high-availability/partials/building-blocks/load-shedding.adoc" />
+
+<#include "/high-availability/partials/building-blocks/sticky-sessions.adoc" />
+
```
I omitted the affinity rules in the CR as the default should work. I explain this more in `single-cluster/concepts.adoc`.

Alexander suggested making the concepts less Kubernetes focused though, so maybe it makes sense to move this information here.

pruivo replied:
I think it makes more sense here, but I'm ok if it isn't :)

ryanemerson commented:
I have addressed all your comments @pruivo except the Kubernetes/Openshift placeholder. I'm going to try to re-work the docs so that the concepts section is more generic and less Kubernetes focused, with Kubernetes details moved to the building blocks sections where possible.

ryanemerson commented on docs/guides/high-availability/introduction.adoc:
```diff
@@ -3,139 +3,56 @@
+== Architectures
 
-== When to use a multi-site setup
+The following Kubernetes based architectures are supported by {project_name}.
```
Done! I have updated all occurrences of Kubernetes or OpenShift so that we use the `{kubernetes}` placeholder. The exceptions being in the test configuration sections and when talking about OpenShift functionality specifically in the building blocks.

ahus1 commented:
Adding #30095 to the list of issues that this will resolve. Let me know if you disagree.

ryanemerson commented on docs/guides/high-availability/single-cluster/concepts.adoc:
```diff
@@ -0,0 +1,161 @@
+| No data loss
+| Less than 30 seconds
+
+| Connectivity {jdgserver_name}
```
The first row relates to tolerating Keycloak pod failures. This row relates to connectivity failures between the Keycloak pods i.e. split-brain.

pruivo commented on docs/guides/high-availability/single-cluster/building-blocks.adoc:
```diff
@@ -0,0 +1,53 @@
+[#single-cluster-blocks-kubernetes]
+== {kubernetes} cluster
+
+*Suggested setup:* Red Hat OpenShift Service on AWS (ROSA) deployed across multiple availability-zones with at least one worker node in each zone.
```
The topic above says `A {kubernetes} cluster consisting of two or more AWS Availability Zones within the same AWS Region.` 
This section seems duplicated.

pruivo commented on docs/guides/high-availability/partials/concepts/threads.adoc:
```diff
@@ -1,33 +1,11 @@
 {project_name} requests, as well as blocking probes, are handled by an executor pool. Depending on the available CPU cores, it has a maximum size of 50 or more threads.
 Threads are created as needed, and will end when no longer needed, so the system will scale up and down automatically.
-{project_name} allows configuring the maximum thread pool size by the link:{links_server_all-config_url}?q=http-pool-max-threads[`http-pool-max-threads`] configuration option. See <@links.ha id="deploy-keycloak-kubernetes" /> for an example.
+{project_name} allows configuring the maximum thread pool size by the link:{links_server_all-config_url}?q=http-pool-max-threads[`http-pool-max-threads`] configuration option. See <@links.ha id="multi-cluster-deploy-keycloak-kubernetes" /> for an example.
```
The last link does not contain an example. Maybe remove it?

pruivo commented on docs/guides/high-availability/single-cluster/deploy-keycloak.adoc:
```diff
@@ -0,0 +1,94 @@
+
+. Install the {project_name} Operator as described in the <@links.operator id="installation" /> {section}.
+
+. Notice the configuration file below contains options relevant for connecting to the Aurora database from <@links.ha id="single-cluster-deploy-aurora" anchor="single-cluster-aurora-connecting" />
```
Double check if the `anchor="single-cluster-aurora-connecting"` exists. The link seems broken, but I haven't built the website to double check.

pruivo commented on docs/guides/high-availability/partials/aurora/aurora-architecture.adoc:
```diff
@@ -0,0 +1,14 @@
+See the https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/CHAP_AuroraOverview.html[AWS Aurora documentation] for more details on the semantics provided by Aurora databases.
+
+This documentation follows AWS best practices and creates a private Aurora database that is not exposed to the Internet.
+To access the database from a ROSA cluster, <<${parent}-establish-peering-connection-with-rosa-cluster,establish a peering connection between the database and the ROSA cluster>>.
```
same here.

pruivo commented on docs/guides/high-availability/multi-cluster/deploy-keycloak-kubernetes.adoc:
```diff
@@ -0,0 +1,54 @@
+* Understanding of a <@links.operator id="basic-deployment" /> of {project_name} with the {project_name} Operator.
+* AWS Aurora database deployed using the <@links.ha id="multi-cluster-deploy-aurora" /> {section}.
+* {jdgserver_name} server deployed using the <@links.ha id="multi-cluster-deploy-infinispan-kubernetes-crossdc" /> {section}.
+* Running {project_name} with OpenJDK 21, which is the default for the containers distributed for {project_name}, as this enabled virtual threads for the JGroups communication.
```
There is not JGroups communication in multi-cluster. This line can be removed :)

pruivo commented on docs/guides/high-availability/multi-cluster/deploy-keycloak-kubernetes.adoc:
```diff
@@ -0,0 +1,54 @@
+
+. Install the {project_name} Operator as described in the <@links.operator id="installation" /> {section}.
+
+. Notice the configuration file below contains options relevant for connecting to the Aurora database from <@links.ha id="multi-cluster-deploy-aurora" anchor="multi-cluster-aurora-connecting" />
```
anchor seems to be broken here too.

pruivo commented on docs/guides/high-availability/partials/building-blocks/load-shedding.adoc:
```diff
@@ -0,0 +1,18 @@
+
+You might consider limiting the value for `http-pool-max-threads` further because multiple concurrent threads will lead to throttling by {kubernetes} once the requested CPU limit is reached.
+
+See the <@links.ha id="multi-cluster-concepts-threads" /> {section} about load shedding for details.
```
`multi-cluster` is hard-coded. Should be `parent` maybe?

pruivo reviewed (commented):
I think I tested all the links :smile:

ahus1 committed:
Update docs/guides/getting-started/getting-started-scaling-and-tuning.adoc

Signed-off-by: Alexander Schwartz <alexander.schwartz@gmx.net>

ahus1 committed:
Update docs/guides/high-availability/multi-cluster/deploy-keycloak-kubernetes.adoc

Signed-off-by: Alexander Schwartz <alexander.schwartz@gmx.net>

ahus1 reviewed (approved):
Thank you for this big refactoring!
